### PR TITLE
Messages refactor

### PR DESCRIPTION
**Pull request by evansiroky**
- create a new class MessageComponent that creates the messages and can be extended
- refactor all classes that utilize messages
- add some flow typing to some components that weren't typed
- add message linting script
- add some missing messages to language files

### CODE REVIEW

**Comment by evansiroky** [comment]
@landonreed The bulk of the changes are in https://github.com/catalogueglobal/datatools-ui/commit/0bb8625f640cc8d9e6ff08a56072a784a5b60947

**Comment by evansiroky** [comment]
@landonreed all the changes are now included in this PR

File: lib/common/util/config.js [hunk 1]
```diff
@@ -16,16 +16,20 @@ export function getGtfsPlusSpec (): Array<GtfsPlusTable> {
   return window.DT_CONFIG.modules.gtfsplus.spec
 }
 
-export function getComponentMessages (componentName: string): {[string]: any} {
-  return (
+export function getComponentMessages (componentName: string): string => string {
```
Comment (landonreed) [inline lib/common/util/config.js]: This function should get a comment about how it is used and what it returns.

File: scripts/lint-messages.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
+// @flow
+
+const fs = require('fs')
+const path = require('path')
+
+const safeLoad = require('js-yaml').safeLoad
+const objectPath = require('object-path')
+
+let atLeastOneErrorFound = false
```
Comment (landonreed) [inline scripts/lint-messages.js]: It seems like it would be a relatively easy refactor to make this track error count.  We would then have a count of errors to report at the end of the lint. We could also create a error and warn wrapper methods that count up the numbers of warnings/errors and also handle logging.

File: scripts/lint-messages.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        // determine what the message path is
+        // if the first character is not ' or ` it's a variable
+        const firstChar = line[messageIndex + 14]
+        if (firstChar === "'") {
+          // isolate string so regex doesn't pick up anything before
+          const message = line.substring(
+            messageIndex + 15,
+            line.indexOf(firstChar, messageIndex + 16)
+          )
+          objectPath.set(foundMessages, `${curClass}.${message}`, true)
```
Comment (landonreed) [inline scripts/lint-messages.js:91]: This section could use an additional comment or two. It's a bit hard to follow.

File: scripts/lint-messages.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+// iterate through files and tally up what was found and what wasn't
+exploreDir('lib')
+
+// analyze differences of messages in language files
+Object.keys(messages).forEach(langFile => {
+  const langMessages = messages[langFile]
+
+  function reconcileObject (obj, pathSoFar) {
```
Comment (landonreed) [inline scripts/lint-messages.js]: Needs comments

Comment (landonreed) [inline scripts/lint-messages.js]: Needs comments

File: scripts/lint-messages.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      const nextPath = pathSoFar ? `${pathSoFar}.${key}` : key
+      if (typeof obj[key] === 'string') {
+        Object.keys(messages)
+          .filter(compareLangFile => compareLangFile !== langFile)
+          .forEach(compareLangFile => {
+            const compareLangMessages = messages[compareLangFile]['components']
+            if (obj[key] === objectPath.get(compareLangMessages, nextPath)) {
+              // FIXME: make this cause an error once we decide to have actual
+              // translations for all languages
+              console.warn(`WARNING: message ${nextPath} in ${langFile} is the same as the message in ${compareLangFile}`)
```
Comment (landonreed) [inline scripts/lint-messages.js]: Could you consolidate these warnings into a single warning per message path? I.e., `WARNING: message example.abc in english.yml is repeated in francais.yml and espanol.yml`?

File: scripts/lint-messages.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  reconcileObject(langMessages['components'])
+})
+
+// analyze what defined messages in the language files were found and which were not
+Object.keys(messages).forEach(langFile => {
+  const langMessages = messages[langFile]
+
+  function reconcileObject (obj, pathSoFar) {
```
Comment (landonreed) [inline scripts/lint-messages.js]: I find it strange that the `reconcileObject` function name is repeated. Is this a convention in test design?

File: scripts/lint-messages.js [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    Object.keys(obj).forEach(key => {
+      const nextPath = pathSoFar ? `${pathSoFar}.${key}` : key
+      if (typeof obj[key] === 'string') {
+        if (!objectPath.get(foundMessages, nextPath)) {
+          // FIXME: make this cause an error once variables and
+          // template strings are accounted for
+          console.warn(`WARNING: message ${nextPath} in ${langFile} not found in any .js files`)
```
Comment (landonreed) [inline scripts/lint-messages.js]: How do we intend to account fully for variables? Is this really giving us valuable information if we're not accounting for variables and template strings? Maybe we should just report the number and location of messages defined with variables and template strings?

**Review by landonreed — CHANGES_REQUESTED** [review]
Looks good. Just needs a few more comments. There are some discussion items on the lint-messages file.

File: scripts/lint-messages.js (hunk 6 above)
Comment (evansiroky) [reply scripts/lint-messages.js]: No, it's just quick copy-pasta.  I'm consolidating some code, so this won't be repeated.

File: scripts/lint-messages.js (hunk 7 above)
Comment (evansiroky) [reply scripts/lint-messages.js]: We should be fully accounting for messages provided as variables and template strings.  I didn't get around to that bit yet.  I was thinking that we could add comments into the js files in a regular format that describe what possibilities of messages are available.  Then, the script would parse the comment.  Something like this:

```javascript
<ControlLabel>{this.messages(`${f}.label`)}</ControlLabel> // lint-message-possibilities: name.label, logoUrl.label
```

**Review by landonreed — APPROVED** [review]
